### DISCUSSION TITLE
Socket Activation

### DISCUSSION THREAD

**Discussion by ghost** [fastify/fastify#4444] [Q&A] [answered]
Just wondering if anyone has any experience with [socket activation](https://github.com/sorccu/node-socket-activation#readme) via systemd (Linux) or launchd (macOS)?

I have not tested the following yet, but I think it will work.

Any thoughts or notes? If I'm correct, you can't pass an `fd` to node's `http.createServer`, but you can pass it to `listen`, so we must patch the returned server's `listen` method... Also, any options passed to `server.listen()` by Fastify are thrown out, except for backlog...

```ts
const fastify = Fastify({
  /**
   * Example to use socket activation, via file descriptor via something like
   * https://github.com/sorccu/node-socket-activation
   */
  serverFactory(handler, opts: FastifyServerOptions) {
    const server = http.createServer(handler);
    // const coreListen = server.listen as net.Server["listen"]; // overload 7
    const coreListen = server.listen as (...args: any[]) => Http.Server;
    Object.assign(server, {
      listen(options: net.ListenOptions, cb?: () => void) {
        const fd = 3; // TODO: Get fd from socket-activation lib...
        if (options.backlog) {
          coreListen.call(server, { fd }, options.backlog, cb);
        } else {
          coreListen.call(server, { fd }, cb);
        }
        return server;
      },
    });
    return server;
  },
  // ...
});
```

**Comment by metcoder95** [comment, marked as answer]
Never tried `socket-activation` before (but thanks for the item to my reading list though). But `fastify` sanitizes the `listen` arguments before calling the `serverFactory` (either custom or the HTTP one from Node). See reference: https://github.com/fastify/fastify/blob/c0686975c9cc72d0f1923428d7fed8afa5d2f1d5/lib/server.js#L305

That being said, at the given state you'll need to pass through the closure of the `listen` method, the `fd` to forward it.
Not sure why `fastify` doesn't support it, but most likely due to security concerns 🙂

**Reply by ghost** [reply]
Thanks for your input! I rewrote my notes from last night's experiment in order to make them more generic.

[Systemd Socket Activation for PWA](https://gist.github.com/waynebloss/63b4a8a326c4ef4e5db24a209eab5882)

- https://gist.github.com/waynebloss/63b4a8a326c4ef4e5db24a209eab5882

**Reply by ghost** [reply]
(In my gist above, I just used a proxy service to forward all HTTP requests instead of opening an `fd`, but I may try it out with the `fd` technique.)

**Reply by metcoder95** [reply]
Sounds interesting and funny. Thanks for the notes, makes a quick sum up of the topic 🙂 
Towards `fastify` that's the only way to bypass the sanitization for passing the `fd` down to the HTTP server creation. Let us know if you were able to make it work 🤞

**Reply by climba03003** [reply]
You can start `fastify` with `fastify.listen({ fd: 3 })`.
The normalization is used to support old multiple arguments that will be deprecated in next major.
